### ISSUE TITLE
Add unlimited categories to the core

### ISSUE THREAD

**Issue by mr-Qberry** [textpattern/textpattern#1121]
Having the ability to add multiple categories to articles and media would be incredibly useful. This would allow content to be displayed in multiple categories rather than just two.

**Comment by Bloke** [comment]
I agree this is useful and has long-been on my list of things to tackle. The main issue stopping me has been backwards compatibility:

* The `textpattern` database table has `category1` and `category2` dedicated columns - expanding this is not sustainable so we need a category-to-content mapping table.
* The other content types only have one column for category name - again, needs to use the mapping table.
* The tags and attributes for things like `<txp:article>` make reference to `category1` and `category2` explicitly. What happens when we allow more?
* Migrating existing category assignments to a new table is definitely doable - ensuring that the `category1` value occupies an ID=1 in the new table, and category2 is in slot number 2 - it's just a hassle on upgrade.
* How to filter by one or more categories from the URL at once, with ALL/ANY support. See smd_tags for possible solutions, if they can be tidied up.
* Category management itself is a pain if you want to maintain lists of categories and keep them in sync across one or more content types. That's more a UX issue separate from unlimited cats, but it's kind of loosely tied with it.

My thinking is a clean departure:

1. Introduce a new table that maps categories to content.
1. Possibly permit txp:category to accept an `id` attribute that would extract the category by number from the current content cat assignment. Its use would be discouraged but it could be used internally (see below).
1. On upgrade, copy existing category assignments for all assets to the mapping table so they are mirrored. Thus `<txp:category1 />` and `<txp:category id="1" />` are interchangeable. Delete old category columns once content is migrated.
1. Introduce a container tag for extracting the categories assigned to content. e.g. `<txp:categories type="article" wraptag="ul" break="li"><txp:category /></txp:categories>` would show a list of all cats assigned to the current article. The `type` would be inferred from context, and overridable if desired. Not sure how this will interact with the existing use of the URL-aware `<txp:category>` though, so we might have to come up with a different naming convention (e.g. `genus` or `subject` or something).
1. Introduce conditional tags such as `<txp:if_categories>` to test if content contains any subject assignment, or test whether it has one or more particular subjects. Bonus marks for ALL/ANY matching. This is essentially `<txp:if_article_category>` without the `number` attribute, and applicable to all content types.
1. Ensure that if a tag uses both new `category` and `categoryN`  attributes that the former are used and the latter are ignored. Internally, we'll silently map this on your behalf.

Once copied over during upgrade, the new table is used to manage mapping cats to content and all existing content category tags are retrofitted to use it. This maintains backwards compatibility and permits growth.

Long term, the category1/2 tags and attributes can be phased out.

We can probably keep the category naming convention, but I fear it is going to get confusing so perhaps a new convention is better for assignments that are actually made to an asset compared to a list of available categories that may be assigned/available for filtering from the URL. We need to be mindful that:

* `<txp:category_list>` iterates over the category structure to build lists of available cats of a particular type.
* `<txp:categories>` (or something) iterates over the subset of categories that are assigned to a particular content asset.
* `<txp:category>` pulls out the current category from inside the `<txp:category_list>` container or the current category in the URL, and *not* the category assigned to a chunk of content. Unless we can find a way to unambiguously dual-purpose it.

Some thought required here.

**Comment by austingovella** [comment]
I create CMS platforms for enterprises that have lots of little sites, and I've noticed common themes around metadata that might be useful here.

1. Sometimes you assign one category, sometimes you assign multiple. And sometimes this changes over time.
2. Sometimes you need the same set of categories to apply to multiple types of content.
3. Some "buckets" of metadata change very frequently, and others change very, very slowly.
4. Sometimes you have hierarchical metadata, sometimes you don't.

If we're thinking big... :-)

- It would be great if site admins could identify the metadata "bucket" they needed.
- The site admin can give that bucket any name they want. Eg: They could call it "category" or "genre" or whatever, so content authors can see the label in the admin.
- The site admin can indicate whether the author can select one or multiple items from that bucket.
- The site admin can indicate whether or not the bucket is required.
- The site admin can identify a default value.
- The site admin can indicate what kind of content the "bucket" applies to: Articles, Images, Files, Links. And the bucket can apply to more than one kind of content. Eg: a bucket called "art era" can apply to both Articles and Images.
- The site admin can indicate whether or not a bucket will contain a flat list or a hierarchical list of metadata.

This is really a lot like custom fields. Maybe there's a path forward where we do NOT mess with either category or custom fields. Instead we create this metadata/custom data functionality and let site admins choose what feature they want to use. They can go with a new system that uses a new tag, or they can continue to use an old feature like custom fields or categories. This would let us keep backward compatibility, as well.

**Comment by Bloke** [comment]
For reference, working draft spec for this feature:

https://docs.google.com/document/d/1ViKpSUwchShkd7B7-m8Gs5zYtJ6Ihc3UHsiPx50uxUQ/edit?usp=sharing

Please comment directly in the spec if there is anything overlooked or if you have better suggestions. Or add comments here.
